### DISCUSSION TITLE
Support all openid-client Client settings via configuration (specifically environment variables)

### DISCUSSION THREAD

**Discussion by illoyd** [directus/directus#20299] [Feature Requests]
### Summary

Currently, the OpenID Auth provider passes most, but not all, properties are passed to the openid-client `Client` constructor.

Allowed properties for openid-client `Client`:
https://github.com/panva/node-openid-client/blob/main/docs/README.md#client
- client_id: <string>
- client_secret: <string>
- id_token_signed_response_alg: <string> Default: 'RS256'
- id_token_encrypted_response_alg: <string>
- id_token_encrypted_response_enc: <string>
- userinfo_signed_response_alg: <string>
- userinfo_encrypted_response_alg: <string>
- userinfo_encrypted_response_enc: <string>
- redirect_uris: string[]
- response_types: string[] Default: '["code"]'
- post_logout_redirect_uris: string[]
- default_max_age: <number>
- require_auth_time: <boolean> Default: 'false'
- request_object_signing_alg: <string>
- request_object_encryption_alg: <string>
- request_object_encryption_enc: <string>
- token_endpoint_auth_method: <string> Default: 'client_secret_basic'
- introspection_endpoint_auth_method: <string> Default: same as token_endpoint_auth_method
- revocation_endpoint_auth_method: <string> Default: same as token_endpoint_auth_method
- token_endpoint_auth_signing_alg: <string>
- introspection_endpoint_auth_signing_alg: <string>
- revocation_endpoint_auth_signing_alg: <string>
- tls_client_certificate_bound_access_tokens: <string>

However, the properties beginning with `id` are not passed to the `Client` as per example below.

### Basic Example

Specifically, we need to set `token_endpoint_auth_method` and `id_token_signed_response_alg`.

The OpenIDAuthDriver supports passing arbitrary environment variables to the openid-client `Client` instance when prefixed with `AUTH_{name}_CLIENT_`.

We expect we can set the environment variable such as `AUTH_{name}_CLIENT_TOKEN_ENDPOINT_AUTH_METHOD` and `AUTH_{name}_CLIENT_ID_TOKEN_SIGNED_RESPONSE_ALG`.

Due to the 2nd parameter of `getConfigFromEnv`, it will filter any key that starts with `AUTH_{name}_CLIENT_ID` and `AUTH_{name}_CLIENT_SECRET`.

This incorrectly removes the `id_*` properties.

### Motivation

Support OpenID Connect advanced settings including different signing algorithms for ID Tokens.

Otherwise, Directus is not truly compliant with OpenID.

### Detailed Design

I propose to enhance `getConfigFromEnv` with smarter options.

Specifically, I would recommend changing the signature to use an `options` parameter, given that there are new options. To support backwards compatability, I recommend the current method name and signature be mapped to a new method.

```
function getConfigFromEnvWithOptions(
	prefix: string,
	options: {
		omitPrefix?: string | string[],
		omitKey?: string | string[],
		type: 'camelcase' | 'underscore' = 'camelcase',
	}
): Record<string, any>
``` 

`omitPrefix` and `type` functions remain the same.

`omitKey` will reject the whole key, rather than performing a prefix check.

The original function can then map to the new method so as to maintain the signature.

In the OpenIDAuthDriver, use the new method with `omitKey` instead of `omitPrefix`.

This should resolve the issue with no breaking impacts on other uses of `getConfigFromEnv`.

### Must Have:
- A Support all properties of openid-client `Client`

### Drawbacks

Potential performance impact when `getConfigFromEnv` maps to `getConfigFromEnvWithOptions`.

Breaking change if the `getConfigFromEnv` signature is changed - so propose to use new method with `options` parameter.

### Alternatives

Directus will not be compliant with OpenID specifications.

### Adoption Strategy

No breaking change, but may consider deprecating the old method signature and replacing in future.

**Comment by overpod** [comment]
I need CLIENT_ID_TOKEN_SIGNED_RESPONSE_ALG for integration with [Logto](https://logto.io/)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
